Make the ambiguous check the same as in runtime

jongleb opened the pull request:
### Description
This PR adds the missing ambiguous checks for `WHERE`, `ORDER BY`, `GROUP BY` and `HAVING` for some cases which were not previously supported

One of these cases (that are reflected in tests) is:

#### DDL:
```sql
CREATE TABLE test21 (id INT, column_a TEXT, column_b BOOL);
CREATE TABLE test22 (id INT, column_d INT);
```

#### Queries

##### First
```sql
SELECT test21.id 
FROM test21 JOIN test22 ON test21.id = test22.id 
WHERE id > 2 ORDER BY id
```

##### Second
```sql
SELECT test21.id 
FROM test21 JOIN test22 ON test21.id = test22.id 
ORDER BY id
```

The first one fails unlike the second one, because `ORDER BY`, `HAVING`, `GROUP BY`  allow have column without explicit referring to source if it's specified in SELECT when ambiguous fields are present.

### Ambiguous Column Rule:

I have empirically tested all cases across three databases: MySQL, PostgreSQL and SQLite. Here are the main tests that are designed to be false positives (i.e., cases that should fail) along with their corresponding Playground links:

[Testing on PostgreSQL](https://www.db-fiddle.com/f/o7ENo2ga8mR3YJfsiyrwzv/6)
[Testing on MySQL](https://www.db-fiddle.com/f/o7ENo2ga8mR3YJfsiyrwzv/7)
[Testing on SQLite](https://www.db-fiddle.com/f/o7ENo2ga8mR3YJfsiyrwzv/5)

### Comparison of versions

##### DDL

```sql
CREATE TABLE test21 (id INT, column_a TEXT, column_b BOOL);
CREATE TABLE test22 (id INT, column_d INT);
CREATE TABLE test2 (id INT, str TEXT);
CREATE TABLE test (id INT, str TEXT, name TEXT);
```

Ambiguous error - 🔴
No errors - 🟢
Warnings - 🟡
Syntax error - ❓

<table>
  <tr>
    <th>Query</th>
    <th>Before this PR <a href="https://github.com/ahrefs/sqlgg/tree/acee7df6cd4c9821ed420fe6e1f7e11fd3847bf9">Branch</a></th>
    <th>Since this PR</th>
    <th>Mysq</th>
    <th>Pgsql</th>
    <th>Sqlite</th>
    <th>Ambiguous reason/expression</th>
    <th>Playground</th>
  </tr>
  <tr>
    <td><pre lang="sql">SELECT test21.id 
FROM test21 
JOIN test22 ON test21.id = test22.id 
WHERE id > 2;</pre></td>
    <td>🟢</td>
    <td>🔴</td>
    <td>🔴</td>
    <td>🔴</td>
    <td>🔴</td>
    <td><code>WHERE</code> expression</td>
    <td><a href="https://www.db-fiddle.com/f/m2hKgSpHr2J2PKmxHeDEg7/1">Link</a></td>
  </tr>
  <tr>
  <tr>
    <td><pre lang="sql">UPDATE test, (SELECT * FROM test2) AS x
SET str = x.str
WHERE test.id = x.id;
</pre></td>
    <td>🟢</td>
    <td>🔴</td>
    <td>🔴</td>
    <td>❓</td>
    <td>❓</td>
    <td><code>SET</code> expression</td>
    <td><a href="https://www.db-fiddle.com/f/rXgv5UT2vLnBNhMWDTvaDH/0">Link</a></td>
  </tr>
<tr>
  <tr>
    <td><pre lang="sql">SELECT *
FROM test21
JOIN test22 on test21.id = test22.id
GROUP BY id;
</pre></td>
    <td>🟢</td>
    <td>🟡 It throws </br> warning .
   But this</br> warning isn't </br> related with <code>GROUP BY</code> !
  </td>
    <td>🔴</td>
    <td>🔴</td>
    <td>🔴</td>
    <td><code>GROUP</code> expression</td>
    <td><a href="https://www.db-fiddle.com/f/7f1gkmZ1Uois1NFhxusP82/0">Link</a></td>
  </tr>
<tr>
  <tr>
    <td><pre lang="sql">SELECT *
FROM test21
JOIN test22 on test21.id = test22.id
ORDER BY id;
</pre></td>
    <td>🟢</td>
    <td>🟡 It throws </br> warning .
   But this</br> warning isn't </br> related with <code>ORDER BY</code> !
  </td>
    <td>🔴</td>
    <td>🔴</td>
    <td>🔴</td>
    <td><code>ORDER</code> expression</td>
    <td><a href="https://www.db-fiddle.com/f/znX9rcqN8exsRkgqcvpLW/0">Link</a></td>
  </tr>
<tr>
  <tr>
    <td><pre lang="sql">SELECT MAX(id) as id2
FROM test21
WHERE id2 > 0;
</pre></td>
    <td>🟢</td>
    <td>🔴</td>
    <td>🔴</td>
    <td>🔴</td>
    <td>🔴</td>
    <td><code>ORDER</code> expression</td>
    <td><a href="https://www.db-fiddle.com/f/ct6Q22GiUcUo4hvkhAbxT9/0">Link</a></td>
  </tr>
</table>

rr0gi commented:
to clarify : this ambiguity rule is something from SQL standard (which?) or database specific (mysql?) or observed empirically?

rr0gi commented on lib/syntax.ml:
```diff
@@ -476,28 +484,24 @@ and eval_select env { columns; from; where; group; having; } =
-     a1.Schema.Source.Attr.attr.name = a2.attr.name && a1.attr.name <> "") in
   (* use schema without aliases here *)
   let p1 = get_params_of_columns env columns in
-  let env = { env with schema = Schema.Join.cross env.schema final_schema |> make_unique } in (* enrich schema in scope with aliases *)
```
enrich comment lost but it is useful

rr0gi commented on lib/syntax.ml:
```diff
@@ -178,7 +185,12 @@ let resolve_aggregations =
+  handle ~is_agg:false 
+  
+let resolve_ambiguous_columns all_schema final_schema = 
+  List.filter (fun s1 ->
```
filter only applies to all_schema - make it more clear

rr0gi replied:
this was not addressed
i mean `let l = List.filter ... all_schema in l @ final_schema`

jongleb replied:
Ah, I thought you meant to me to name it more clearly. Made more clear ( it to be not to looked that `@` applies to final_shema and all_schema before filter applies to all_schema)

rr0gi commented on lib/syntax.ml:
```diff
@@ -178,7 +185,12 @@ let resolve_aggregations =
-  handle ~is_agg:false  
+  handle ~is_agg:false 
+  
+let resolve_ambiguous_columns all_schema final_schema = 
```
the name resolve_ is confusing - it is not resolving anything, it is just adding alias columns to schema

jongleb commented:
> to clarify : this ambiguity rule is something from SQL standard (which?) or database specific (mysql?) or observed empirically?

Added this part to the description

rr0gi commented:
i still hard time wrapping my head around the change
the links to db-fiddle showcase different queries than the ones in the table
the table is missing the column what is the behaviour of these queries in the databases
the  second row in the table i don't see anything ambiguous

jongleb commented:
> i still hard time wrapping my head around the change
> the links to db-fiddle showcase different queries than the ones in the table

I also added the second table with queries from the db-fiddle example. The first table demonstrates examples from given from tests.

> the table is missing the column what is the behaviour of these queries in the databases

I added it to the both tables.

> the second row in the table i don't see anything ambiguous

Yes, and as we see it work since this PR

<img width="1016" alt="second_row" src="https://github.com/user-attachments/assets/b37fd982-9725-4100-867c-2d09f6d1e034">

rr0gi commented:
> Yes, and as we see it work since this PR

it works before this PR as well, there is no ambiguity there

```
[..]
  module List = struct
    let select_4 db  callback =
      let invoke_callback stmt =
        callback
          ~id:(T.get_column_Int_nullable stmt 0)
          ~id0:(T.get_column_Int_nullable stmt 1)
      in
      let r_acc = ref [] in
      IO.(>>=) (T.select db ("SELECT test21.id, test22.id\n\
FROM test21 \n\
JOIN test22 ON test21.id = test22.id") T.no_params (fun x -> r_acc := invoke_callback x :: !r_acc))
      (fun () -> IO.return (List.rev !r_acc))

  end (* module List *)
end (* module Sqlgg *)
Warning: this SQL statement will produce rowset with duplicate column names:
SELECT test21.id, test22.id
FROM test21 
JOIN test22 ON test21.id = test22.id
```
it gives the warning on duplicate column names which is valid, but it generates code alright

rr0gi commented:
```
SELECT id, id
FROM test21 
JOIN test22 ON test21.id = test22.id 
GROUP BY id;
```
this example is orthogonal to this PR, it is duplicate columns in the result set again, does not test ambiguity.

rr0gi commented:
so i mean PR is useful and i guess i understand which cases are handled, but the description in PR is still very confusing (or just plain wrong), need to fix because this is the only documentation that we have now.

rr0gi commented:
the queries in the test cases in the PR make sense, idu why the ones you pick for the documentation tables are different %)

rr0gi commented on src/test.ml:
```diff
@@ -358,6 +359,72 @@ let test_agg_nullable = [
+  tt "select test21.id from test21 join test22 on test21.id = test22.id order by id" [
+    attr' ~nullability:(Nullable) "id" Int;
+  ] [];
+  wrong "select test21.id from test21 join test22 on test21.id = test22.id where id > 2 order by id";
```
please put comment from the PR here why this one doesn't fail

jongleb replied:
But it fails. I call here the "wrong" function. I means assert fail. Added comment that says that wrong does assert fail

rr0gi replied:
i meant comment this difference in behaviour between where vs order by

jongleb replied:
Added comment

rr0gi commented on src/test.ml:
```diff
@@ -358,6 +359,72 @@ let test_agg_nullable = [
+    attr' ~nullability:(Nullable) "id" Int;
+    attr' ~nullability:(Nullable) "id" Int;
+  ] [];
+  wrong "select id, id from test21 join test22 on test21.id = test22.id group by id";
```
```suggestion
  wrong "select id as id1, id as id2 from test21 join test22 on test21.id = test22.id group by id";
```

jongleb replied:
Kept it 
```sql
wrong "select id, id from test21 join test22 on test21.id = test22.id group by id";
```
Since this call "wrong" function, that does assert fail.
And added your example too
```sql
wrong "select id as id1, id as id2 from test21 join test22 on test21.id = test22.id group by id";
```

rr0gi commented on src/test.ml:
```diff
@@ -358,6 +359,72 @@ let test_agg_nullable = [
+  wrong "select * from test21 join test22 on test21.id = test22.id group by id" ;
+  tt "CREATE TABLE test23 (id INT)" [] [];
+  tt "CREATE TABLE test24 (id INT)" [] [];
+  wrong "select * from foo join bar on foo.id = bar.id order by id";
```
this is same test as above (line 390)?

jongleb replied:
Yes, deleted

jongleb replied:
Yes, deleted.

rr0gi commented on src/test.ml:
```diff
@@ -358,6 +359,72 @@ let test_agg_nullable = [
+  tt "CREATE TABLE test24 (id INT)" [] [];
+  wrong "select * from foo join bar on foo.id = bar.id order by id";
+  wrong "select * from foo join bar on foo.id";
+  tt "SELECT test21.id AS id, test22.id AS id FROM test21 JOIN test22 ON test21.id = test22.id" [
```
```suggestion
  tt "SELECT test21.id AS id1, test22.id AS id2 FROM test21 JOIN test22 ON test21.id = test22.id" [
```
nobody should write two AS id so lets not confuse ppl

jongleb replied:
Fixed

rr0gi commented on lib/syntax.ml:
```diff
@@ -107,10 +107,17 @@ let exists_grouping columns =
+let make_unique =
+  List.unique ~cmp:(fun a1 a2 ->
+      (* Check if alias has same name as column *)
+      (List.length a2.Schema.Source.Attr.sources = 0
+      (* Check if columns are from the same table (source)  *)
+      || (List.length a1.sources = List.length a2.sources && List.for_all2 ( = ) a1.sources a2.sources))
```
```suggestion
     (* Check if columns are from the same table (source)  *)
      (a2.Schema.Source.Attr.sources = [] || a1.sources = a2.sources)
```
less confusing indentation and simplify

rr0gi replied:
why is a2.sources=[] check is needed? it means if a2.sources is empty and a1.sources is not empty then result will be true, ie equality is not symmetric? :thinking:

jongleb replied:
> why is a2.sources=[] check is needed? it means if a2.sources is empty and a1.sources is not empty then result will be true, ie equality is not symmetric? 🤔

```sql 
SELECT COUNT(column_a) as column_a FROM test21 WHERE column_a = @column_a
```
This is when we alias a column to the same name as a column in selected tables.
Maybe this is a good pattern to do that at all , but 
1) sql supports it
2) we have this examples in our code

rr0gi replied:
put this in code comment
i am still not convinced why `make_unique a b` is allowed to be different from `make_unique b a`

jongleb replied:
> i am still not convinced why make_unique a b is allowed to be different from make_unique b a

Fixed

rr0gi commented on lib/syntax.ml:
```diff
@@ -178,7 +185,12 @@ let resolve_aggregations =
+  
+let update_schema_with_aliases all_schema final_schema = 
+  List.filter (fun s1 ->
+    List.for_all (fun s2 -> s2.Schema.Source.Attr.attr.name <> s1.Schema.Source.Attr.attr.name) final_schema
```
what about unnamed columns btw, `"" <> ""` is it important here?

jongleb replied:
It's function applied to `Column` expression, but unnamed. Like example above but without  alias.
```sql
SELECT COUNT(column_a) FROM test21 WHERE column_a = @column_a
```
If we have more unnamed function applied to columns expressiono as columns,

```sql
SELECT COUNT(column_a), AVG(column_a) FROM test21 WHERE column_a = @column_a
```

the case of uniqueness will always be matched here.
we give names at a later stage already during code generation for and only for the arguments in the function for mapping results

rr0gi commented on lib/syntax.ml:
```diff
@@ -107,10 +107,17 @@ let exists_grouping columns =
   List.exists (function Expr (e,_) -> is_grouping e | All | AllOf _ -> false) columns
 
 (* all columns from tables, without duplicates *)
-(* FIXME check type of duplicates *)
```
this fixme disappeared, i think it is important, at some point we want to warn if there are columns with same name but different types

jongleb replied:
Okay, I'll return it. I just thought my PR put a dot to the ambiguity check, and this is why I deleted it.

jongleb commented:
> ```
> SELECT id, id
> FROM test21 
> JOIN test22 ON test21.id = test22.id 
> GROUP BY id;
> ```
> this example is orthogonal to this PR, it is duplicate columns in the result set again, does not test ambiguity.

Removed

jongleb commented:
> it gives the warning on duplicate column names which is valid, but it generates code alright

First of all, you are right, sorry. I didn't see this warning when I wrote tests. Because this warning happens on the later stages, but tests are checked for the earlier stage (parsing + param type inferring). I would say since my PR it happens at the early stage (since in runtime it's runtime error, not warning) + GROUP BY, ORDER BY, WHERE, HAVING statements weren't cheked at all

and make this query valid

```sql
UPDATE test, (SELECT * FROM test2) AS x 
SET str = x.str 
WHERE test.id = x.id;
``` 


, since  

```ocaml
          if not (Sql.Schema.is_unique stmt.schema) then
            Printf.eprintf "Warning: this SQL statement will produce rowset with duplicate column names:\n%s\n" (fst sql);
``` 

this warning checks only schema. I Added "Main distinguishing requests" chapter to the description of the PR. Or does it make senee to delete tables above and to make a new one ?

UPD: I updated table, warnings aren't related to WHERE/GROUP BY/ even SET and the rest any statements not from schema (code with warning checks schema)

rr0gi commented on lib/syntax.ml:
```diff
@@ -108,9 +108,15 @@ let exists_grouping columns =
+      (* Check if alias has same name as column OR if columns are from the same table (source)  *)
+      (a2.Schema.Source.Attr.sources = [] || a1.Schema.Source.Attr.sources = a2.sources)
+      && a1.attr.name = a2.attr.name
+      (* Check if the tested column is unnamed (not a column) and unaliased *)
```
btw this comment is not correct (and related to my previous comment about this function becoming non-symmetrical)
what it is testing is that both columns are not unnamed (because of comparison above), there is no such thing as "tested column"

jongleb replied:
Fixed, now is symmetrical. 

```sql
  tt "SELECT COUNT(column_a) as column_a FROM test23 WHERE column_a = @column_a" [
    (* COUNT(column_a :: Text) :: Int *)
    attr' "column_a" Int;
  ] [
    named "column_a" Text;
  ];
```
I was keeping it: `a2.Schema.Source.Attr.sources = []` , but realized that WHERE doesn't support aliases. Removed now

jongleb replied:
Updated table with example

jongleb replied:
```sql
SELECT MAX(id) as id2
FROM test21
WHERE id2 > 0;
GROUP BY column_b;
```
This isn't valid

rr0gi commented on lib/syntax.ml:
```diff
@@ -178,7 +184,13 @@ let resolve_aggregations =
+  handle ~is_agg:false 
+  
+let update_schema_with_aliases all_schema final_schema = 
+  let applied = List.filter (fun s1 ->
+    List.for_all (fun s2 -> s2.Schema.Source.Attr.attr.name <> s1.Schema.Source.Attr.attr.name) final_schema
+  ) all_schema in 
```
```suggestion
  let applied = all_schema |> List.filter (fun s1 -> List.for_all Schema.Source.Attr.(fun s2 -> s2.attr.name <> s1.attr.name) final_schema) in 
```

jongleb replied:
Done

rr0gi commented on src/test.ml:
```diff
@@ -207,7 +208,7 @@ let test_coalesce = [
 ]
 
-let test_coalesce = [
+let test_coalesce2 = [
```
please rebase, this was fixed in master

jongleb replied:
Done

rr0gi reviewed (changes requested):
so `make_unique` has me concerned, otherwise ok

rr0gi commented on lib/syntax.ml:
```diff
@@ -124,9 +124,15 @@ let exists_grouping columns =
+    (* Check if columns are from the same table (source)  *)
+    a1.Schema.Source.Attr.sources = a2.sources
+    && a1.attr.name = a2.attr.name
+    (* Check if columns are unnamed (for example a function is applied to a column) and unaliased *)
```
this comment sounds wrong, because it actually checks that columns _are named_

jongleb replied:
```SELECT MAX(col) FROM smth```
in this case, it checks `if name <> ""`, because `if name = ""` we skip it (consider them as unnamed) because we can have more than one field such that, so yes, you are right.

rr0gi commented on src/test.ml:
```diff
@@ -489,6 +490,85 @@ let cte_possible_rec_non_shared_select_only = [
+  tt "CREATE TABLE test24 (id INT, column_d INT)" [] [];
+  wrong "select id from test23 join test24 on test23.id = test24.id order by id";
+  (* The difference between this example, and the same but with WHERE (following "wrong" fn) is
+     sql engine uses those columns that were mentioned in the SELECT statement, what he
```
```suggestion
     sql engine uses those columns that were mentioned in the SELECT statement, while it doesn't do that for WHERE
```
